src/device-state.coffee

```diff
@@ -0,0 +1,512 @@
+							d.localId ?= null
+							d.is_managed_by ?= null
+							d.lock_expiry_date ?= null
+							return d
+						devicesForState = _.map devicesForThisApp, (d) ->
+							dev = {
+								uuid: d.uuid
+								name: d.name
+								apps: {}
+							}
+							dev.apps[d.appId] = {
+								config: tryParseObj(d.targetConfig)
+								environment: tryParseObj(d.targetEnvironment)
+							}
+						legacyTarget.dependent.devices = legacyTarget.dependent.devices.concat(devicesForState)
+						@db.models('dependentDevice').insert(devices)
+				.catch (err) ->
+					console.error("Ignoring legacy dependent app #{app.imageId} due to #{err}")
+		.then =>
+			@setTarget(legacyTarget)
+		.then =>
+			@config.set({ initialConfigSaved: 'true' })
+
+	init: ->
+		@config.getMany([
+			'logsChannelSecret', 'pubnub', 'offlineMode', 'loggingEnabled', 'initialConfigSaved',
+			'listenPort', 'apiSecret', 'osVersion', 'osVariant', 'version', 'provisioned',
+			'resinApiEndpoint', 'connectivityCheckEnabled'
+		])
+		.then (conf) =>
+			@logger.init({
+				pubnub: conf.pubnub
+				channel: "device-#{conf.logsChannelSecret}-logs"
+				offlineMode: conf.offlineMode
+				enable: conf.loggingEnabled
+			})
+			.then =>
+				@config.on 'change', (changedConfig) =>
+					@logger.enable(changedConfig.loggingEnabled) if changedConfig.loggingEnabled?
+					@reportCurrentState(api_secret: changedConfig.apiSecret) if changedConfig.apiSecret?
+			.then =>
+				@applications.init()
+			.then =>
+				if !validation.checkTruthy(conf.initialConfigSaved)
+					@saveInitialConfig()
+			.then =>
+				@initNetworkChecks(conf)
+				console.log('Reporting initial state, supervisor version and API info')
+				@reportCurrentState(
+					api_port: conf.listenPort
+					api_secret: conf.apiSecret
+					os_version: conf.osVersion
+					os_variant: conf.osVariant
+					supervisor_version: conf.version
+					provisioning_progress: null
+					provisioning_state: ''
+					logs_channel: conf.logsChannelSecret
+					update_failed: false
+					update_pending: false
+					update_downloaded: false
+				)
+			.then =>
+				@loadTargetFromFile() if !conf.provisioned
+			.then =>
+				@triggerApplyTarget()
+
+	initNetworkChecks: ({ resinApiEndpoint, connectivityCheckEnabled }) =>
+		network.startConnectivityCheck(resinApiEndpoint, connectivityCheckEnabled)
+		@config.on 'change', (changedConfig) ->
+			network.enableConnectivityCheck(changedConfig.connectivityCheckEnabled) if changedConfig.connectivityCheckEnabled?
+		console.log('Starting periodic check for IP addresses')
+		network.startIPAddressUpdate (addresses) =>
+			@reportCurrentState(
+				ip_address: addresses.join(' ')
+			)
+		, @config.constants.ipAddressUpdateInterval
+
+	saveInitialConfig: =>
+		@deviceConfig.getCurrent()
+		.then (devConf) =>
+			@deviceConfig.setTarget(devConf)
+		.then =>
+			@config.set({ initialConfigSaved: 'true' })
+
+	emitAsync: (ev, args...) =>
+		setImmediate => @emit(ev, args...)
+
+	readLockTarget: =>
```

Comment:
This will expose the release callback to the caller. Change these functions to accept the function to be run under the lock and handle locking and releasing internally

Reply:
Done, here and in the update lock calls in ApplicationManager